### ISSUE TITLE
Is it necessary to add more efficient search indexes such as HNSW?

### ISSUE THREAD

**Issue by ycw66** [FastCodeAI/vectorDB#8]

**Comment by stencilman** [comment]
Hi @ycw66, it will be nice to have, but perhaps only is useful when you are storing a lot of data.

**Comment by stencilman** [comment]
@DhavalTaunk08 @parthbasole99 what do you guys think? @ycw66 did you want to contribute this to dvdb?

**Comment by davidmigloz** [comment]
The current algorithm brute-forces similarity across all stored vectors, which has a complexity of `O(vector_dimensionality * num_vectors)`. It may work fine for a few thousand documents, but it will start to struggle with more or when using larger embedding vectors (e.g. OpenAI's `text-embedding-3-large` which has 3072 dimensions).

**Comment by stencilman** [comment]
yes @davidmigloz you are right, we only have a couple of hundred. We will definitely need a better algorithm when someone needs to use it on a larger dataset.

**Comment by DhavalTaunk08** [comment]
I think for on-device applications people won't be storing huge amount of data. I guess it is more useful for vector-databases which are used for server side applications.

**Comment by ailabs-software** [comment]
@DhavalTaunk08 First, thanks for putting this together. It was great to see such a simple implementation as I am in unfamiliar territory.

But to the original question, what if we were to use dart:ffi to interop with hnswlib?

Would this be a worthwhile weekend experiment to try eliminating the performance impact of linear search?

Of course, this would mean DVDB wouldn't be able to run in the browser (but this doesn't look like a goal of the project anyway).

Perhaps DVDB could have a shared interface with two implementation options, a basic implementation for small projects & a more optimised one.

To add context for my motivation, I am running Dart on the server side (and client side) of a large enterprise web application.

**Comment by davidmigloz** [comment]
@ailabs-software  If you need more efficient and scalable embeddings DBs, you can take a look to:
- [ObjectBox Vector Search](https://objectbox.io/the-first-on-device-vector-database-objectbox-4-0/)
- [sqlite3.dart](https://github.com/simolus3/sqlite3.dart) + [sqlite-vec](https://github.com/asg017/sqlite-vec) ([example](https://github.com/rodydavis/flutter_sqlite_document_search))
